### ISSUE TITLE
zippapp/ macOS: .DS_Store files should be excluded from the generated archives

### ISSUE THREAD

**Issue by TurtleWilly** [python/cpython#117829]
# Bug report

### Bug description:

When generating a zipapp under macOS then random ".DS_Store" files get included into the generated archive.

e.g.
```
$ python3.12 -m zipapp --output cfv.pyz --compress --python /usr/local/bin/python3 zipapptempdir
$ unzip -l cfv.pyz 
Archive:  cfv.pyz 
  Length      Date    Time    Name
---------  ---------- -----   ----
     6148  04-13-2024 00:17   .DS_Store
      186  04-13-2024 00:20   __main__.py
        0  04-13-2024 00:17   cfv/
     6148  04-13-2024 00:17   cfv/.DS_Store
        …
---------                     -------
   137783                     20 files
```

Those files quickly get regenerated by Finder, so it's hard to entirely avoid them getting added to the archive by trying to pre-delete them before executing zipapp. .DS_Store files can add significantly to the size of the archive, possible have privacy implications (e.g. leaked spotlight comments), and serve zero purpose for the python app.

### CPython versions tested on:

3.12

### Operating systems tested on:

macOS

### Linked PRs
* gh-120537

**Comment by Buggem** [comment]
I do not agree. I think by default they should be excluded but there should be an option to include them.

**Comment by gaogaotiantian** [comment]
Quietly ignoring a specific file on MacOS does not seem like the desirable behavior to me. Currently the bahavior of `zipapp` is "pack everything", which is easy to understand. Changing it to "pack everything except for files with name `.DS_Store`" is kind of weird. There are other commonly generated files, `__pycache__` or `.coverage` for example. Who should be the person to decide what's on the list and what's not?

To achieve your goal, the Python API `zipapp.create_archive` provides a way to do it. You can use the argument `filter` to filter the paths. It should be easy to create your own script to generate the zipapp file with customized filtering. I don't believe changing the default behavior is the correct direction.

I can, somehow, imagine a new option to ignore all hidden files, which is probably more generic to users.

**Comment by Buggem** [comment]
> Quietly ignoring a specific file on MacOS does not seem like the desirable behavior to me. Currently the bahavior of `zipapp` is "pack everything", which is easy to understand. Changing it to "pack everything except for files with name `.DS_Store`" is kind of weird. There are other commonly generated files, `__pycache__` or `.coverage` for example. Who should be the person to decide what's on the list and what's not?
> 
> To achieve your goal, the Python API `zipapp.create_archive` provides a way to do it. You can use the argument `filter` to filter the paths. It should be easy to create your own script to generate the zipapp file with customized filtering. I don't believe changing the default behavior is the correct direction.
> 
> I can, somehow, imagine a new option to ignore all hidden files, which is probably more generic to users.
> 
> And this is definitely not a bug, it's a feature request, so I'll change the lables.

agree 100%

**Comment by ronaldoussoren** [comment]
> Quietly ignoring a specific file on MacOS does not seem like the desirable behavior to me. Currently the bahavior of `zipapp` is "pack everything", which is easy to understand. Changing it to "pack everything except for files with name `.DS_Store`" is kind of weird. There are other commonly generated files, `__pycache__` or `.coverage` for example. Who should be the person to decide what's on the list and what's not?

Adding a filter option to the zipapp command line interface could be useful, but as you write the hard bit is which name should be included in the default filter.

Two options (which could both be added)
- `python -m zipapp --filter GLOB ...`: Multi-use argument for specifying which names should be excluded (user needs to determine useful filters themselves)
- `python -m zipapp --filter-junk ..`:  Filter some commonly seen files that should generally not be included, such as `.DS_Store`, `__pycache__` and backup files (`*~`, `._*.swp`, ...). 

The second option is a nicer user interface for users, but is harder to get right (both naming and which filters to include).

Currently the only API implements something similar to the second option is `dircmp`, other APIs try to avoid the temptation to guess ;-).

**Comment by TurtleWilly** [comment]
The problem with the .DS_Store files (unlike other temporary junk mentioned here) is that they can come back randomly and quickly during the time the zippapp is created. E.g. I do run a cleanup before already after I noticed the issue with .DS_Store files ballooning the sizes of the generated archives:

```bash
find zipapptempdir -type f -name '*.py[co]' -delete -o -type d -name __pycache__ -delete -o -type f -name '.DS_Store' -delete && \
  python3.12 -m zipapp --output s3cmd.pyz --compress --python /usr/local/bin/python3.12 zipapptempdir
```

And in one case a .DS_Store file still managed to sneak into the archive. It's very hard to make this work reliable from a single user account. Maybe killing off Finder before, and then relaunching it afterwards or some other awkward kludge?

I'm reporting about the the zipapp app/module itself (as used in `python3 -m zipapp` from a user perspective) and I'm perfectly fine with some `--filter REGEX` or `--reject GLOB` option, if that's more suitable than just default filtering out the `.DS_Store` files. 👍

I would consider those files getting added as a privacy/security/ information leak issue (hence a bug report), e.g. when I plan to distribute generated zipapps to other users: they really don't need my .DS_Store files.

**Comment by gaogaotiantian** [comment]
If we plan to add a command line option to filter, I think we should avoid the name `filter`, unless we can make it comparable to the `filter` argument for `create_archive`. The `filter` argument is a function that returns a boolean indicating whether the path should be accepted. The only way I can think of to make the option `filter` work is for it to pass in a regular expression that provides a pattern for all accepted paths (not declined, or it will have opposite meaning to the argument).

A better way might be `--include` and `--exclude` which is also commonly used to provide both allow-list and deny-list.

I'm a bit surprised that Finder will generate the file in the background (when it's not opened in the OS) though.

**Comment by VietThan** [comment]
@TurtleWilly  @gaogaotiantian @Buggem I've made a draft pull request based on @gaogaotiantian's direction. Before I add tests, I'm wondering if you all have any comments or feedback for me. Thank you.

**Comment by TurtleWilly** [comment]
@VietThan: Looks good from what I can tell. 👍

**Comment by VietThan** [comment]
Pull request https://github.com/python/cpython/pull/120537 is ready to review

**Comment by VietThan** [comment]
Ping @aisk @ronaldoussoren @gaogaotiantian @Buggem @TurtleWilly , the pull request https://github.com/python/cpython/pull/120537 is ready for review.

**Comment by VietThan** [comment]
Ping @aisk @ronaldoussoren @gaogaotiantian , the pull request #120537 is ready for review.

**Comment by TurtleWilly** [comment]
@VietThan: Looks all very good to me. I'd say lets merge. Not sure how this process works here exactly or what else I could do to help out except giving thumbs up everywhere? :)

One minor thing: in the help text I simply would use `unzip -l myapp.pyz` to streamline, no need to extract first and and the use ls afterwards. But this is entirely minor and not related to the actual code, of course.

**Comment by VietThan** [comment]
Yeah I'm not sure either @TurtleWilly , seems like best is to just keep pinging @aisk @ronaldoussoren and/or @gaogaotiantian to review and merge the PR.

**Comment by VietThan** [comment]
Hi @pfmoore , could you review this PR when you have time?

**Comment by pfmoore** [comment]
You should not keep pinging maintainers. The PR has unresolved review comments (and I've just added some more). It's up to you to address the comments, and once you have done so, the PR will be marked for re-review without you needing to ping anyone.

Furthermore, I don't consider this behaviour a bug in the zipapp CLI. It's not up to zipapp to deal with the MacOS behaviour of arbitrarily dumping system files in user directories (especially if, as is claimed here, those files contain security sensitive data). I'm willing to consider a feature request to add filtering capabilities to the command line interface, but the functionality is already available via the `create_archive` API, and so I don't see this as an urgent requirement. Furthemore, I'd prefer to keep the CLI simple, and direct people to the `create_archive` API (probably called from a project-specific utility script) when their needs go beyond the basic "bundle up a local directory" use case. So I'm at best lukewarm on the feature request.

I haven't seen any compelling examples of other (legitimate) use cases where filtering might be required in the CLI. It would be helpful if someone could give examples of where this might be useful in real-world situations (beyond working round the MacOS misbehaviour that's already been mentioned).

**Comment by VietThan** [comment]
My apologies. I didn't know that would work and I wasn't familiar with the GitHub Pull Request interface. Will rework the PR this weekend.
